File: 3700-number-of-zigzag-arrays-ii/solution.py
MOD = 10**9 + 7

def multiply(A, B):
    size = len(A)
    C = [[0] * size for _ in range(size)]
    for i in range(size):
        for k in range(size):
            if A[i][k] == 0: continue
            for j in range(size):
                C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD
    return C

def power(A, p):
    size = len(A)
    res = [[0] * size for _ in range(size)]
    for i in range(size): res[i][i] = 1
    while p > 0:
        if p % 2 == 1: res = multiply(res, A)
        A = multiply(A, A)
        p //= 2
    return res

class Solution:
    def countZigZag(self, n: int, l: int, r: int) -> int:
        K = r - l + 1
        size = 2 * K
        T = [[0] * size for _ in range(size)]
        
        for v in range(K):
            for next_v in range(K):
                if v == next_v: continue
                if v < next_v: T[v][K + next_v] = 1
                else: T[K + v][next_v] = 1
        
        if n == 2: return K * (K - 1)
        
        Tn = power(T, n - 2)
        v2 = [0] * size
        for i in range(K):
            for j in range(K):
                if i < j: v2[K + j] += 1
                elif i > j: v2[j] += 1
        
        ans = 0
        for i in range(size):
            for j in range(size):
                ans = (ans + v2[i] * Tn[i][j]) % MOD
        return ans

File: 3700-number-of-zigzag-arrays-ii/solution.cpp
#include <vector>

using namespace std;

long long MOD = 1e9 + 7;

typedef vector<vector<long long>> Matrix;

Matrix multiply(const Matrix& A, const Matrix& B, int size) {
    Matrix C(size, vector<long long>(size, 0));
    for (int i = 0; i < size; ++i)
        for (int k = 0; k < size; ++k)
            for (int j = 0; j < size; ++j)
                C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
    return C;
}

Matrix power(Matrix A, long long p, int size) {
    Matrix res(size, vector<long long>(size, 0));
    for (int i = 0; i < size; ++i) res[i][i] = 1;
    while (p > 0) {
        if (p & 1) res = multiply(res, A, size);
        A = multiply(A, A, size);
        p >>= 1;
    }
    return res;
}

class Solution {
public:
    int countZigZag(int n, int l, int r) {
        int K = r - l + 1;
        int size = 2 * K;
        Matrix T(size, vector<long long>(size, 0));

        // State indices: 0 to K-1 (value), direction 0 (prev was up/next must be down), 
        // direction 1 (prev was down/next must be up).
        // Actual index = direction * K + value
        for (int v = 0; v < K; ++v) {
            for (int next_v = 0; next_v < K; ++next_v) {
                if (v == next_v) continue;
                if (v < next_v) T[v][K + next_v] = 1;      // Up step: from v to next_v
                else T[K + v][next_v] = 1;                // Down step: from v to next_v
            }
        }

        Matrix Tn = power(T, n - 2, size);
        
        vector<long long> v2(size, 0);
        for (int i = 0; i < K; ++i) {
            for (int j = 0; j < K; ++j) {
                if (i < j) v2[K + j] = (v2[K + j] + 1) % MOD;
                else if (i > j) v2[j] = (v2[j] + 1) % MOD;
            }
        }

        long long ans = 0;
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                ans = (ans + v2[i] * Tn[i][j]) % MOD;
            }
        }
        return (int)ans;
    }
};